cpp/src/arrow/stl_iterator.h

```diff
@@ -128,6 +131,215 @@ class ArrayIterator {
+    } else {
+      iterators_list_[current_chunk_index_]++;
+    }
+    return *this;
+  }
+  ChunkedArrayIterator& operator--() {
+    --index_;
+    if (iterators_list_[current_chunk_index_].index()) {
+      iterators_list_[current_chunk_index_]--;
+    } else {
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      current_chunk_index_--;
+      if (!static_cast<int64_t>(
+              chunked_array_->chunk(static_cast<int>(current_chunk_index_))->length())) {
+        current_chunk_index_ = GetChunkIndex(index_);
+      }
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      iterators_list_[current_chunk_index_] +=
+          static_cast<int64_t>(
+              chunked_array_->chunk(static_cast<int>(current_chunk_index_))->length()) -
+          1;
+    }
+    return *this;
+  }
+
+  ChunkedArrayIterator operator++(int) {
+    ChunkedArrayIterator tmp(*this);
+    ++*this;
+    return tmp;
+  }
+  ChunkedArrayIterator operator--(int) {
+    ChunkedArrayIterator tmp(*this);
+    --*this;
+    return tmp;
+  }
+
+  // Arithmetic
+  difference_type operator-(const ChunkedArrayIterator& other) const {
+    return index_ - other.index_;
+  }
+  ChunkedArrayIterator operator+(difference_type n) const {
+    return ChunkedArrayIterator(*chunked_array_, index_ + n);
+  }
+  ChunkedArrayIterator operator-(difference_type n) const {
+    return ChunkedArrayIterator(*chunked_array_, index_ - n);
+  }
+  friend inline ChunkedArrayIterator operator+(difference_type diff,
+                                               const ChunkedArrayIterator& other) {
+    return ChunkedArrayIterator(*other.chunked_array_, diff + other.index_);
+  }
+  friend inline ChunkedArrayIterator operator-(difference_type diff,
+                                               const ChunkedArrayIterator& other) {
+    return ChunkedArrayIterator(*other.chunked_array_, diff - other.index_);
+  }
+  ChunkedArrayIterator& operator+=(difference_type n) {
+    if (n < static_cast<int64_t>(
+                chunked_array_->chunk(static_cast<int>(current_chunk_index_))->length()) -
+                iterators_list_[current_chunk_index_].index()) {
+      index_ += n;
+      iterators_list_[current_chunk_index_] += n;
+      return *this;
+    } else {
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      index_ += n;
+      InitializeComponents(*this, true);
+      return *this;
+    }
+  }
+  ChunkedArrayIterator& operator-=(difference_type n) {
+    if (n <= iterators_list_[current_chunk_index_].index()) {
+      index_ -= n;
+      iterators_list_[current_chunk_index_] -= n;
+      return *this;
+    } else {
+      iterators_list_[current_chunk_index_] -=
+          iterators_list_[current_chunk_index_].index();
+      index_ -= n;
+      InitializeComponents(*this, true);
+      return *this;
+    }
+  }
```

Comment:
This can be simplified to have a single impl because `operator-=(n) == operator+=(-n)`